[FEATURE] seqan3::sam_record

marehr opened the pull request:
Part of https://github.com/seqan/product_backlog/issues/289

This implements
* sam_record
* adds member to sam_record, see names https://github.com/seqan/product_backlog/issues/289
* and does some infrastructure

eaasna commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<2>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The mate pair information given as a std::tuple of reference name, offset and template length.
```
```suggestion
    //!\brief The mapping quality of the seqan3::field::seq alignment, usually a Phred-scaled score. 
```

marehr replied:
nice thank you!

eaasna commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+    //!\}
+
+    //!\brief The identifier, usually a string.
+    decltype(auto) id() &&
```
Could you please explain why there are 4 different assignments for each field? The slight differences in syntax are not clear to me

marehr replied:
Yes, we want to mimic the way `std::get` behaves, and it will return 

* `id_type &&` for `std::get<id>(sam_record &&)`
* `id_type &` for `std::get<id>(sam_record &)`
* `id_type const &&` for `std::get<id>(sam_record const &&)`
* `id_type const &` for `std::get<id>(sam_record const &)`

We unfortunately need 4 overloads for this depending on whether the input has the type modifier `&&`, `&`, `const &&`, `const &`. The return type, here `decltype(auto)`, has the corresponding type modifier within it.

This is a special case, and this is not the norm, so you should not write 4 accessors functions if you write code yourself. In most cases it is fine just to copy the data without having a [ref/const-qualified function](https://en.cppreference.com/w/cpp/language/member_functions#ref-qualified_member_functions).

eaasna replied:
Thanks for the explanation!

eaasna reviewed (changes requested):
Thanks a lot for the effort!

Just a small correction and a question

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::seq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The qualities, usually in phred-score notation.
```
```suggestion
    //!\brief The qualities, usually in Phred score notation.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::qual>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief Sequence (SEQ) relative start position (0-based), unsigned value.
```
```suggestion
    //!\brief Sequence (SEQ) relative start position (0-based), unsigned value.
```

Mhm, should it link to the `sequence()` function instead of `SEQ`? Now we suddenly use SAM field notation where we used the SeqAn entities before - and we do not link to the SAM notation always either...

marehr replied:
```suggestion
    //!\brief Sequence (seqan3::sam_record::sequence) relative start position (0-based), unsigned value.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::alignment>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The identifier of the (reference) sequence that SEQ was aligned to.
```
Same here

marehr replied:
```suggestion
    //!\brief The identifier of the (reference) sequence that seqan3::sam_file::sequence was aligned to.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief Sequence (REF_SEQ) relative start position (0-based), unsigned value.
```
SAM does not have `REF_SEQ`

marehr replied:
```suggestion
    //!\brief Sequence (seqan3::sam_record::reference_sequence) relative start position (0-based), unsigned value.
```

marehr replied:
you are right and we don't set it currently in sam. You can infer this from the id-sequence-map.

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::flag>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The mate pair information with the reference name.
```
```suggestion
    //!\brief The identifier of the (reference) sequence of the mate.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<0>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The mate pair information with the position.
```
```suggestion
    //!\brief Sequence (SEQ) relative start position (0-based) of the mate.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<1>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The mate pair information with the template length.
```
```suggestion
    //!\brief The observed template length.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<2>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The mapping quality of the seqan3::field::seq alignment, usually a Phred-scaled score. 
```
```suggestion
    //!\brief The mapping quality of the alignment, usually a Phred-scaled score. 
```

I find this redundant...

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::mapq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The cigar vector (std::vector<seqan3::cigar>) representing the alignment in SAM/BAM format.
```
```suggestion
    //!\brief The cigar vector (std::vector<seqan3::cigar>) representing the alignment.
```

There's nothing else now :)

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::cigar>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The optional tags in the SAM format, stored in a dictionary.
```
```suggestion
    //!\brief The optional tags in the SAM format, stored in a seqan3::sam_tag_dictionary.
```

eseiler reviewed (changes requested):
Just documentation stuff, I'm not sure about all the phrasings either

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::offset>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The (pairwise) alignment stored in an seqan3::alignment object.
```
```suggestion
    //!\brief The (pairwise) alignment stored in an object that models seqan3::detail::pairwise_alignment.
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<0>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief Sequence (SEQ) relative start position (0-based) of the mate.
```
```suggestion
    //!\brief Sequence relative start position (0-based) of the mate.
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+    using base_t::base_t;
+    //!\}
+
+    //!\brief The identifier, usually a string.
```
```suggestion
    //!\brief The identifier, usually a string. (QNAME)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::id>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The "sequence", usually a range of nucleotides or amino acids.
```
```suggestion
    //!\brief The "sequence", usually a range of nucleotides or amino acids. (SEQ)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::seq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The qualities, usually in Phred score notation.
```
```suggestion
    //!\brief The qualities, usually in Phred score notation. (QUAL)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::ref_id>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The (reference) "sequence" information, usually a range of nucleotides or amino acids.
+    decltype(auto) reference_sequence() &&
+    {
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t &&>(*this));
+    }
+    //!\copydoc seqan3::sam_record::reference_sequence
+    decltype(auto) reference_sequence() const &&
+    {
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t const &&>(*this));
+    }
+    //!\copydoc seqan3::sam_record::reference_sequence
+    decltype(auto) reference_sequence() &
+    {
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t &>(*this));
+    }
+    //!\copydoc seqan3::sam_record::reference_sequence
+    decltype(auto) reference_sequence() const &
+    {
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t const &>(*this));
+    }
```
```suggestion
    //!\brief The (reference) "sequence" information, usually a range of nucleotides or amino acids. (Currently not implemented!)
    decltype(auto) reference_sequence() = delete;
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::header_ptr>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The alignment flag (bit information), `uint16_t` value.
```
```suggestion
    //!\brief The alignment flag (bit information), `uint16_t` value. (SAM Column ID: FLAG)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::flag>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The identifier of the (reference) sequence of the mate.
```
```suggestion
    /*!\brief The identifier of the (reference) sequence of the mate. (SAM Column ID: RNEXT)
     * \details
     * If `RNEXT` is `=` it returns the same as seqan3::sam_record::reference_id.
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<1>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The observed template length.
```
```suggestion
    //!\brief The observed template length. (SAM Column ID: TLEN)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return std::get<2>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief The mapping quality of the alignment, usually a Phred-scaled score. 
```
```suggestion
    //!\brief The mapping quality of the alignment, usually a Phred-scaled score. (SAM Column ID: MAPQ)
```

marehr commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,415 @@
+        return get_impl(field_constant<seqan3::field::mapq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The cigar vector (std::vector<seqan3::cigar>) representing the alignment.
```
```suggestion
    //!\brief The cigar vector (std::vector<seqan3::cigar>) representing the alignment. (SAM Column ID: QUAL)
```

marehr commented:
@eseiler I added a couple of suggestions, can you go over them and edit and then apply them? You are completely right that seqan3::field::ref_seq aka seqan3::sam_record::reference_sequence isn't populated at all, and I "deleted" it for now.

eseiler commented:
> @eseiler I added a couple of suggestions, can you go over them and edit and then apply them? You are completely right that `seqan3::field::ref_seq` aka `seqan3::sam_record::reference_sequence` isn't populated at all, and I "deleted" it for now.

I thought you can construct the `alignment_file_input` with additional reference information, and then it works?

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,428 @@
+        return get_impl(field_constant<seqan3::field::qual>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    /*!\brief Sequence (seqan3::sam_record::sequence) relative start position (0-based), unsigned value. (SAM Column ID: POS)
+    *
+     * \details
+     * The position is the length of the soft-clipping at the start of the seqan3::sam_record::cigar_sequence if a soft-clipping is present, 0 otherwise.
+     */
```
```suggestion
    /*!\brief Sequence (seqan3::sam_record::sequence) relative start position (0-based), unsigned value.
     *        (SAM Column ID: POS)
     *
     * \details
     *
     * The position is the length of the soft-clipping at the start of the seqan3::sam_record::cigar_sequence if a 
     * soft-clipping is present and `0` otherwise.
     */
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,428 @@
+        return get_impl(field_constant<seqan3::field::offset>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    /*!\brief Sequence (seqan3::sam_record::sequence) relative start position (0-based), unsigned value.  
+     *        (SAM Column ID: POS)
+     *
+     * \details
+     *
+     * The position is the length of the soft-clipping at the start of the seqan3::sam_record::cigar_sequence if a 
+     * soft-clipping is present and `0` otherwise.
+     */
```
```suggestion
    //!\brief The (pairwise) alignment stored in an object that models seqan3::detail::pairwise_alignment.
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,428 @@
+        return get_impl(field_constant<seqan3::field::alignment>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief The identifier of the (reference) sequence that seqan3::sam_record::sequence was aligned to. (SAM Column ID: RNAME)
```
```suggestion
    /*!\brief The identifier of the (reference) sequence that seqan3::sam_record::sequence was aligned to. 
     *        (SAM Column ID: RNAME)
     */
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,428 @@
+        return get_impl(field_constant<seqan3::field::ref_seq>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    //!\brief (Reference) Sequence (seqan3::sam_record::reference_sequence) relative start position (0-based), unsigned value. (SAM Column ID: POS)
```
```suggestion
    /*!\brief (Reference) Sequence (seqan3::sam_record::reference_sequence) relative start position (0-based), 
     *        unsigned value. (SAM Column ID: POS)
     */
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,428 @@
+        return get_impl(field_constant<seqan3::field::flag>{}, static_cast<tuple_base_t const &>(*this));
+    }
+
+    /*!\brief The identifier of the (reference) sequence of the mate. (SAM Column ID: RNEXT)
+     * \details
+     * If `RNEXT` is `=` it returns the same as seqan3::sam_record::reference_id.
+    decltype(auto) mate_reference_id() &&
+    {
```
```suggestion
    /*!\brief The identifier of the (reference) sequence of the mate. (SAM Column ID: RNEXT)
     *
     * \details
     *
     * If `RNEXT` is `=`, it returns the same as seqan3::sam_record::reference_id.
     */
    decltype(auto) mate_reference_id() &&
    {
```

eseiler commented on include/seqan3/io/alignment_file/record.hpp:
```diff
@@ -0,0 +1,431 @@
+        return std::get<0>(get_impl(field_constant<seqan3::field::mate>{}, static_cast<tuple_base_t const &>(*this)));
+    }
+
+    //!\brief (Reference) Sequence relative start position (0-based) of the mate.
```
```suggestion
    //!\brief (Reference) Sequence relative start position (0-based) of the mate. (SAM Column ID: PNEXT)
```

marehr commented:
> > @eseiler I added a couple of suggestions, can you go over them and edit and then apply them? You are completely right that `seqan3::field::ref_seq` aka `seqan3::sam_record::reference_sequence` isn't populated at all, and I "deleted" it for now.
> 
> I thought you can construct the `alignment_file_input` with additional reference information, and then it works?

You are right, I can add this if you want. I thought of this Pr as a reintroduction of existing, working functionality, but this would be more like a new "feature". I'd  want to do it in a second PR, because I need to add new test cases for it.

eseiler commented:
> > > @eseiler I added a couple of suggestions, can you go over them and edit and then apply them? You are completely right that `seqan3::field::ref_seq` aka `seqan3::sam_record::reference_sequence` isn't populated at all, and I "deleted" it for now.
> > 
> > 
> > I thought you can construct the `alignment_file_input` with additional reference information, and then it works?
> 
> You are right, I can add this if you want. I thought of this Pr as a reintroduction of existing, working functionality, but this would be more like a new "feature". I'd want to do it in a second PR, because I need to add new test cases for it.

Ah, so it works for the `alignment_file_record`, but not for the `sam_record`? This is ok for me.

Do you want to rebase?

marehr commented:
> > > > @eseiler I added a couple of suggestions, can you go over them and edit and then apply them? You are completely right that `seqan3::field::ref_seq` aka `seqan3::sam_record::reference_sequence` isn't populated at all, and I "deleted" it for now.
> > > 
> > > 
> > > I thought you can construct the `alignment_file_input` with additional reference information, and then it works?
> > 
> > 
> > You are right, I can add this if you want. I thought of this Pr as a reintroduction of existing, working functionality, but this would be more like a new "feature". I'd want to do it in a second PR, because I need to add new test cases for it.
> 
> Ah, so it works for the `alignment_file_record`, but not for the `sam_record`? This is ok for me.
> 
> Do you want to rebase?

I added a backlog issue https://github.com/seqan/product_backlog/issues/292
